### PR TITLE
Increased ingest speed

### PR DESCRIPTION
**Pull request by olcbean**
This change gives a 4x speedup.
* increase the 'carry' size to process more than 127 bytes at a time
* decrease the num of queues, 
* add ability to push the same piece of memory through the layers 

While the code could be made even smarter, pushing this version is a good starting point.

### CODE REVIEW

**Comment by ribasushi** [comment]
Thank you @olcbean, this is just excellent! 🚀 

@anjor this is a big part of what you were looking for, preliminary tests seem to be pretty great. Please take over review/merging when time permits.

File: commp.go [hunk 1]
```diff
@@ -123,28 +130,33 @@ func (cp *Calc) Digest() (commP []byte, paddedPieceSize uint64, err error) {
...
-		if len(cp.carry) < 127 {
-			cp.carry = append(cp.carry, make([]byte, 127-len(cp.carry))...)
+	if len(cp.buffer) > 0 {
+		if mod := len(cp.buffer) % quadPayload; mod != 0 {
+			cp.buffer = append(cp.buffer, make([]byte, quadPayload-mod)...)
+		}
+		for len(cp.buffer) > 0 {
+			// FIXME: there is a smarter way to do this instead of 127-at-a-time,
+			// but that's for another PR
+			cp.digestQuads(cp.buffer[:127])
```
Comment (anjor) [inline commp.go:149]: is this 127 logically the same as quadPayload, or is it a different number?

**Comment by anjor** [comment]
Thanks for the work @olcbean . 

Do you have some numbers/tests with the 4x improvement? I ran a couple of tests on a 5GB payload and a 8GB payload and not seeing much difference (`stream-commp` is this branch, `stream-commp-master` is the master branch):

```
anjor@seven car-generation-testing $ time cat 8G-payload.bin.car | ~/repos/filecoin-project/go-fil-commp-hashhash/cmd/stream-commp/stream-commp -d

CommPCid: baga6ea4seaqhdtg3kstze5qogwjx5r5bjsegoesqfvk3iyxyh3tg5c7vjzm56hi
Payload:          8592870123 bytes
Unpadded piece:  17045651456 bytes
Padded piece:    17179869184 bytes

real	0m44.529s
user	1m18.730s
sys	0m26.671s

anjor@seven car-generation-testing $ time cat 8G-payload.bin.car | ~/repos/filecoin-project/go-fil-commp-hashhash/cmd/stream-commp/stream-commp-master -d

CommPCid: baga6ea4seaqhdtg3kstze5qogwjx5r5bjsegoesqfvk3iyxyh3tg5c7vjzm56hi
Payload:          8592870123 bytes
Unpadded piece:  17045651456 bytes
Padded piece:    17179869184 bytes

real	0m42.815s
user	1m15.518s
sys	0m22.792s
```

```
anjor@seven car-generation-testing $ time cat 5gb-payload.bin.car | ~/repos/filecoin-project/go-fil-commp-hashhash/cmd/stream-commp/stream-commp -d

CommPCid: baga6ea4seaqla4fdrxtk2l3yd74jg23y3weqlfh2p3fkfueuwlf4oou7qhgkmea
Payload:          5454458475 bytes
Unpadded piece:   8522825728 bytes
Padded piece:     8589934592 bytes

real	0m27.250s
user	0m47.973s
sys	0m14.888s

anjor@seven car-generation-testing $ time cat 5gb-payload.bin.car | ~/repos/filecoin-project/go-fil-commp-hashhash/cmd/stream-commp/stream-commp-master -d

CommPCid: baga6ea4seaqla4fdrxtk2l3yd74jg23y3weqlfh2p3fkfueuwlf4oou7qhgkmea
Payload:          5454458475 bytes
Unpadded piece:   8522825728 bytes
Padded piece:     8589934592 bytes

real	0m26.943s
user	0m47.914s
sys	0m14.235s
```

File: commp.go [hunk 2]
```diff
@@ -50,42 +50,49 @@ const MaxPiecePayload = MaxPieceSize / 128 * 127
...
 	stackedNulPadding [MaxLayers][]byte
 )
 
 // initialize the nul padding stack (cheap to do upfront, just MaxLayers loops)
 func init() {
 	h := shaPool.Get().(hash.Hash)
-	defer shaPool.Put(h)
 
-	stackedNulPadding[0] = make([]byte, 32)
+	stackedNulPadding[0] = make([]byte, commpDigestSize)
```
Comment (anjor) [inline commp.go:69]: // not a review, just capturing for posterity

`sha256simd.Size` is also 32 -- https://github.com/minio/sha256-simd/blob/v1.0.0/sha256.go#L29

**Comment by anjor** [comment]
Capturing some performance tests :
```
anjor@seven car-generation-testing $ time cat 8G-payload.bin.car | ~/repos/filecoin-project/go-fil-commp-hashhash/cmd/stream-commp/stream-commp-master

CommPCid: baga6ea4seaqhdtg3kstze5qogwjx5r5bjsegoesqfvk3iyxyh3tg5c7vjzm56hi
Payload:          8592870123 bytes
Unpadded piece:  17045651456 bytes
Padded piece:    17179869184 bytes

CARv1 detected in stream:
Blocks:     32960
Roots:          1
    1: bafkqaaa


real	0m45.085s
user	1m18.703s
sys	0m24.709s
anjor@seven car-generation-testing $ time cat 8G-payload.bin.car | ~/repos/filecoin-project/go-fil-commp-hashhash/cmd/stream-commp/stream-commp

CommPCid: baga6ea4seaqhdtg3kstze5qogwjx5r5bjsegoesqfvk3iyxyh3tg5c7vjzm56hi
Payload:          8592870123 bytes
Unpadded piece:  17045651456 bytes
Padded piece:    17179869184 bytes

CARv1 detected in stream:
Blocks:     32960
Roots:          1
    1: bafkqaaa


real	0m13.486s
user	0m31.589s
sys	0m6.037s
```

```
anjor@seven car-generation-testing $ time cat 4G-payload.bin.car | ~/repos/filecoin-project/go-fil-commp-hashhash/cmd/stream-commp/stream-commp-master

CommPCid: baga6ea4seaqlisjwvo5wqyw2artgldlmhnrdpyw2i3xzsvu7tvw552lseentmmy
Payload:          4296435046 bytes
Unpadded piece:   8522825728 bytes
Padded piece:     8589934592 bytes

CARv1 detected in stream:
Blocks:     16480
Roots:          1
    1: bafkqaaa


real	0m21.932s
user	0m39.209s
sys	0m11.829s

anjor@seven car-generation-testing $ time cat 4G-payload.bin.car | ~/repos/filecoin-project/go-fil-commp-hashhash/cmd/stream-commp/stream-commp

CommPCid: baga6ea4seaqlisjwvo5wqyw2artgldlmhnrdpyw2i3xzsvu7tvw552lseentmmy
Payload:          4296435046 bytes
Unpadded piece:   8522825728 bytes
Padded piece:     8589934592 bytes

CARv1 detected in stream:
Blocks:     16480
Roots:          1
    1: bafkqaaa


real	0m6.548s
user	0m15.804s
sys	0m3.107s
```

File: commp.go [hunk 3]
```diff
@@ -280,60 +287,71 @@ func (cp *Calc) addLayer(myIdx uint) {
...
+func (cp *Calc) hashSlab254(layerIdx uint, slab []byte) {
 	h := shaPool.Get().(hash.Hash)
-	h.Reset()
-	h.Write(half1ToOverwrite)
-	h.Write(half2)
-	d := h.Sum(half1ToOverwrite[:0]) // callers expect we will reuse-reduce-recycle
-	d[31] &= 0x3F
-	out <- d
+
+	stride := 1 << (5 + layerIdx)
```
Comment (anjor) [inline commp.go:347]: sorry, i am not quite following the index + 5 choice here -- mind walking me through it?

**Comment by anjor** [comment]
@olcbean overall looks good. I left a few comments -- mind addressing those? We can then get it merged quickly!

File: commp.go (hunk 1 above)
Comment (olcbean) [reply commp.go:149]: yes, it is, left as is - it is in the `FIXME` ;)

File: commp.go [hunk 4]
```diff
@@ -161,115 +173,110 @@ func (cp *Calc) Digest() (commP []byte, paddedPieceSize uint64, err error) {
...
+		cp.buffer = append(cp.buffer, input...)
+		return len(input), nil
+	}
 
-		// super short Write - just carry it
-		if carrySize+inputSize < 127 {
-			cp.carry = append(cp.carry, input...)
-			return inputSize, nil
-		}
+	totalInputBytes := len(input)
```
[the remark this answers is not in the document]
Comment (olcbean) [reply commp.go:206]: `input` is being changed ( resliced ) on L210, L218 so we are recording its original size

File: commp.go (hunk 3 above)
Comment (olcbean) [reply commp.go:347]: The leaves are 32 Bytes ( 1 << 5 )
On layer0 we hash every 2 neighbouring leaves - pos0, pos1, pos2, pos3, .... and the results are written in pos0, ..., pos2, ... 
On the next layer pos0 and pos2 are hashed and the result is stored in pos0 -> stride 64, etc

**Review by anjor — APPROVED** [review]

**Pull request merged**